### PR TITLE
feat(events): provide the ability to register global hammer.js events

### LINKED ISSUE #7308: [Feature request] Global key event filtering
When having a component that wants to listen to global key events, using `(body:keyup.escape)` crashes with an `EXCEPTION: not implemented`, whereas `(body:keyup)` works fine. Is this actually being worked on, or just not in the roadmap at all?

Here is my use case: I'm trying to build a modal component, and I want to dismiss it when the user presses escape. And any listener tied to the modal itself never triggers, so I need to listen to the keyup on `body`.

``` typescript
@Component({
    selector: "my-modal",
    host: {
        "(body:keyup.escape)": "dismiss()"
    }
    ...
})
export class MyModal {
    @Input("MyModalOpen") open = false;
    @Output("MyModalOpenChange") closed = new EventEmitter<boolean>();
    ...
}
```

The only solution I see right now is catch every keyup and filter myself, which feels dirty. I'm open to any alternative solutions. :smile: 
Thanks for your help.

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/7308 and https://github.com/angular/angular/issues/12566

Original pr https://github.com/angular/angular/pull/10596 had too many merge conflicts so I decided to create a new one.

### CODE REVIEW

**Comment by vicb** [comment]
Please split this into 
- a refactor commit,
- a fix commit

**Comment by vicb** [comment]
& also add a test for the fix. Thanks !

**Comment by DzmitryShylovich** [comment]
@vicb do u have an idea how I can test it? There's a check for global Hammer object https://github.com/angular/angular/blob/master/modules/%40angular/platform-browser/src/dom/events/hammer_gestures.ts#L56 and my test failed with `Error: Hammer.js is not loaded, can not bind swipe event`

**Comment by vicb** [comment]
IIUC the pb is that `addGlobalEventListener` was not implemented. You can instantiate the plugim and `expect(() => addGlobalEventListener(...)).not.toThrow()`, you probably need a spy for `addEventListener`.

If you feel like, it would also be nice to merge h_common and h_gestures, the split was only required for Dart and makes no sense now.

Thanks !

**Comment by DzmitryShylovich** [comment]
@vicb done

File: modules/@angular/platform-browser/src/dom/events/hammer_gestures.ts [hunk 1]
```diff
@@ -66,9 +106,7 @@ export class HammerGesturesPlugin extends HammerGesturesPluginCommon {
     return zone.runOutsideAngular(() => {
       // Creating the manager bind events, must be done outside of angular
       const mc = this._config.buildHammer(element);
-      const callback = function(eventObj: HammerInput) {
-        zone.runGuarded(function() { handler(eventObj); });
-      };
+      const callback = (eventObj: HammerInput) => { zone.runGuarded(() => handler(eventObj)); };
```
Comment (vicb) [inline modules/@angular/platform-browser/src/dom/events/hammer_gestures.ts]: pls revert this one (we don't want to bind the context here).

**Comment by vicb** [comment]
Great !!!

One small nit before I can merge

**Comment by DzmitryShylovich** [comment]
done

**Pull request closed without merging**